cms/0017-media-tracking.md

```diff
@@ -0,0 +1,99 @@
+
+For this RFC we will focus on Media tracking.
+
+The goal of this RFC is to detail the MVP (Minimum Viable Product) for an initial release of Media tracking in the CMS and shed light on any issues or pitfalls with the approach described below. In the future, we will add further support for tracking Umbraco item relations.
+
+## Motivation
+
+There is currently no way to tell if an Umbraco item is being referenced. This makes cleanup a difficult process since deleting an item that is referenced can have an unintentional impact on content and potentially break functionality on a site. Adding tracking and overview of Umbraco item usage would help increase confidence when working in the backoffice as users have a chance to either avoid deleting an item or remove/change referenced items before deleting, mitigating unexpected outcomes. It will enable users and developers to be proactive instead of reactive when working with Umbraco items.
+
+## Detailed Design
+
+### Property Editors
+
+The fundamental change in Umbraco to support this is to update the Property Editor APIs so that a Property Editor itself returns the UDIs that it is referencing. This will occur whenever an Umbraco content item (Content, Media, Member) is saved that uses Property Editors. When Umbraco saves the data, it will iterate over all of its property types and ask the corresponding Property Editor to return a list of UDIs. 
+
+The method could be something along the lines of:
+
+```cs
+IEnumerable<Udi> GetReferences(TContent content, Property property);
+```
+
+This means the power of resolving related items is in the hands of the Property Editor since it knows exactly what data it is storing and how to process it.
+
+The property editors that will need to be updated to use this method are:
+
+* Content Picker
+* Media Picker
+* Member Picker
+* MNTP
+* Rich Text Editor
+* Grid
+* Multi URL Picker
+* Nested Content
+* Markdown editor
+
+### Relations
+
+When an Umbraco content item (Content, Media, Member) is saved and all of the UDI references have been resolved from the Property Editors, we will store these relations in the Umbraco Relations database table using the Umbraco Relations Service APIs. For content, we will __only__ be tracking relations for "Published" content _(see unresolved issues below for more info)_.
+
+The relations Service APIs may need to be enhanced to properly support Umbraco media item tracking and reporting detailed in this RFC.
+
+_NOTE:_ We need to investigate how Umbraco Deploy is handling relations since we don't want to reinvent anything. We will either leverage what Umbraco Deploy is doing or Umbraco  Deploy will leverage this implementation.
```

Comment:
Deploy can handle relations in different ways. It is configurable (see: https://our.umbraco.com/documentation/Umbraco-Cloud/Deployment/Deploy-Settings/#relationtypes). I believe it will default to `Weak` if nothing is specified, which means the relation will be applied/stored if both items exist in the final state on the destination where the items are deployed to - it will not make dependencies based on relations when in using this mode. It might be an idea to leverage Deploy to include the specific relation types we end up using for reference tracking, as strong dependencies to ensure Deploy picks them up as actual dependencies (I guess that is what would make sense to do for this).

Relations are transferred _alongside_ the actual entity artifacts being deployed and aren't a part of the actual artifacts and doesn't have an actual artifact representation itself either.